### PR TITLE
[ROCm] Add Triton block-sparse attention backend for CacheBlend

### PR DESCRIPTION
**Pull request by andyluo7**
## Summary

Add `LMCTritonSparseBackend` as a drop-in replacement for `LMCFlashInferSparseBackend` that works on ROCm (AMD GPUs) via Triton kernels. This enables CacheBlend non-prefix KV cache reuse on AMD MI300X and MI355X GPUs without requiring flashinfer.

### New files
- **`triton_kernels/block_sparse_attention.py`** — 3 Triton kernels:
  - `_block_sparse_attn_fwd_kernel`: block-sparse attention with CSR sparse indices + LSE output
  - `_causal_prefill_attention_kernel`: standard causal prefill (replaces flashinfer `single_prefill`)
  - `_merge_attention_kernel`: LSE-based output blending for CacheBlend
- **`triton_sparse.py`** — `LMCTritonSparseBackend(AttentionInterface)` implementation

### Modified files
- **`metadata.py`** — Add `LMCTritonSparseMetadata` with CSR index conversion from block masks
- **`utils.py`** — Auto-route to Triton backend on ROCm or when flashinfer is unavailable; lazy imports for vLLM 0.18+ compatibility
- **`lookup_client/lmcache_lookup_client.py`** — Convert `ConstantList` to plain `list` before msgpack encoding (vLLM 0.18+ compat)
- **`compute/models/base.py`** — Fallback embedding lookup chain for vLLM 0.18+ (`embed_input_ids` / `model.embed_tokens`)

### Kernel unit tests (MI355X, gfx950)
```
Cosine similarity: 1.000000 ✅
Max output diff: 0.000488 ✅
Max LSE diff: 0.000001 ✅
Merge max diff: 0.000031 ✅
All tests passed! ✅
```

### E2E CacheBlend (vLLM 0.18.0+rocm700, Qwen2.5-1.5B-Instruct)

**MI300X (gfx942):**
```
Cold (req1):  0.54s
Blend (req2): 2.17s  (Triton JIT compilation)
Blend (req3): 0.53s
✅ E2E test completed without errors!
```

**MI355X (gfx950):**
```
Cold (req1):  0.46s
Blend (req2): 1.36s  (Triton JIT compilation)
Blend (req3): 0.33s
✅ E2E test completed without errors!
```

## Notes
- The Triton backend also works on CUDA as a flashinfer-free alternative
- `utils.py` auto-detects ROCm and routes accordingly; no user configuration needed
- E2E testing requires the `gpu_worker.py` model tracker patch documented in `examples/blend_kv_v1/README.md` (related: #2306)
- Pin-count warnings during blending are a known upstream issue (#2954)

---

> [!NOTE]
> **Medium Risk**
> Adds a new Triton-based block-sparse attention implementation and changes backend selection logic, which can affect numerical correctness and performance across CUDA/ROCm paths. Also updates vLLM integration points (embeddings + lookup serialization), which may impact runtime compatibility if edge cases aren’t covered.
> 
> **Overview**
> Enables CacheBlend sparse attention on ROCm (and as a flashinfer-free fallback) by introducing a new `LMCTritonSparseBackend` powered by Triton block-sparse attention kernels.
> 
> Adds `LMCTritonSparseMetadata` that converts CacheBlend’s selected `top_indices` into CSR (`block_indices`/`block_indptr`) for the Triton kernel, and updates attention backend inference to auto-select Triton on ROCm or when `flashinfer` is unavailable.
> 
> Improves vLLM compatibility by adding embedding-method fallbacks in `compute_layer` and ensuring lookup requests serialize `token_ids` as a plain Python `list`; includes a new ROCm-focused E2E test exercising CacheBlend with the Triton sparse path.
> 
> <sup>Reviewed by [Cursor Bugbot](https://cursor.com/bugbot) for commit 13650fbe370393fbe9c35c1b9e014f4399846aa5. Bugbot is set up for automated code reviews on this repo. Configure [here](https://www.cursor.com/dashboard/bugbot).</sup>

### CODE REVIEW

**Review by sammshen — APPROVED** [review]
LGTM!

File: lmcache/v1/compute/attention/triton_kernels/__init__.py [hunk 1]
```diff
@@ -0,0 +1,6 @@
+# SPDX-License-Identifier: Apache-2.0
+from .block_sparse_attention import (
+    block_sparse_attention,
+    causal_prefill_attention,
+    merge_attention_outputs,
```
[the remark this answers is not in the document]
Comment (andyluo7) [reply lmcache/v1/compute/attention/triton_kernels/__init__.py:6]: Intentional — `merge_attention_outputs` and `_lse` are part of the public API for the LSE-based two-path blending pattern (sparse-cached blocks + recomputed blocks). The current CacheBlend flow in `base.py` uses a single combined KV pass, but the merge kernel is needed for the more advanced split-path flow where sparse attention and full recompute run separately and their outputs are blended via log-sum-exp weighting. Keeping it avoids a follow-up PR when that path is enabled.

Similarly, `_lse` is stored on the metadata so downstream consumers (e.g., a future `blend_layer` variant) can access it without re-running the sparse attention.

**Comment by ApostaC** [comment]
@YaoJiayi Please take a look at this PR, thanks!

**Comment by YaoJiayi** [comment]
@andyluo7 The code looks good to me but it seems that we are rebuilding the wheels? Can we use existing attention libraries (e.g., flexattention)?

**Comment by andyluo7** [comment]
Good question! We evaluated FlexAttention (`torch.nn.attention.flex_attention`) but chose Triton for several reasons:

1. **ROCm compatibility** — FlexAttention relies on `torch.compile` to generate fused kernels, and the ROCm backend for `torch.compile` does not fully support FlexAttention's block-sparse patterns yet. Since the primary motivation for this PR is ROCm/AMD GPU support, FlexAttention would not solve the problem today.

2. **Sparse format mismatch** — CacheBlend produces CSR-indexed block-sparse masks (`block_indices`/`block_indptr`), while FlexAttention uses its own `BlockMask` format with `kv_num_blocks`/`kv_indices` layout. The conversion adds complexity without clear benefit.

3. **Minimal surface area** — The Triton implementation is three small kernels (~400 lines total) with no dependencies beyond Triton itself. This keeps the ROCm support self-contained and easy to maintain.

That said, FlexAttention is the right long-term direction once its ROCm support matures. The kernel API surface here is intentionally small, so migrating to FlexAttention in a future PR would be straightforward — the `LMCTritonSparseBackend` interface stays the same, only the kernel implementation changes.

**Comment by andyluo7** [comment]
@ApostaC  @YaoJiayi, any further comment?

File: lmcache/v1/compute/attention/metadata.py [hunk 2]
```diff
@@ -83,3 +83,100 @@ def update_from_top_indices(self, top_indices: torch.Tensor):
...
+    num_q_blocks = block_mask.shape[0]
+    nnz_per_row = block_mask.sum(dim=1)
+    block_indptr = torch.zeros(num_q_blocks + 1, dtype=torch.int32, device=device)
+    torch.cumsum(nnz_per_row, dim=0, out=block_indptr[1:])
+    total_nnz = block_indptr[-1].item()
+    if total_nnz == 0:
+        return (
+            torch.zeros(0, dtype=torch.int32, device=device),
+            block_indptr,
+        )
```
Comment (deng451e) [inline lmcache/v1/compute/attention/metadata.py]: Small nit: .item() here likely forces a CPU/GPU sync, which could lag the pipeline. Could we skip the early return and let torch.where handle the empty case naturally? It returns an empty tensor on an all-False mask, so the result should be the same.

Comment (andyluo7) [reply lmcache/v1/compute/attention/metadata.py]: Good catch — you're right that `.item()` forces a sync. The early return was a defensive guard but `torch.where` handles the empty case correctly already. I'll remove the `.item()` check and let it flow through naturally.

File: lmcache/v1/compute/attention/triton_kernels/block_sparse_attention.py [hunk 3]
```diff
@@ -0,0 +1,418 @@
...
+    # Query block range
+    q_start = q_block_idx * BLOCK_M
+    q_offs = q_start + tl.arange(0, BLOCK_M)
+    q_mask = q_offs < seq_len_q
+
+    # Load the sparse KV block range for this query block
+    # Guard against grid blocks exceeding CSR rows (partial block case)
+    if q_block_idx >= num_csr_rows:
+        # No CSR entry for this block — write zeros and -inf LSE
+        o_ptrs = O_ptr + q_offs[:, None] * stride_om + head_idx * stride_oh + d_offs[None, :] * stride_od
```
[the remark this answers is not in the document]
Comment (andyluo7) [reply lmcache/v1/compute/attention/triton_kernels/block_sparse_attention.py]: Fixed — `d_offs` is now defined before the early-return branch (line 63), so it's always available.

File: lmcache/v1/compute/attention/triton_kernels/block_sparse_attention.py [hunk 4]
```diff
@@ -0,0 +1,391 @@
...
+    """
+    M, H, D = query.shape
+    N = key.shape[0]
+
+    output = torch.empty_like(query)
+    lse = torch.empty(M, H, dtype=torch.float32, device=query.device)
+
+    num_q_blocks = (M + block_size - 1) // block_size
+
+    grid = (num_q_blocks, H)
```
[the remark this answers is not in the document]
Comment (andyluo7) [reply lmcache/v1/compute/attention/triton_kernels/block_sparse_attention.py:425]: Fixed — the kernel now receives `num_csr_rows` and guards with `if q_block_idx >= num_csr_rows: return` (writes zeros + -inf LSE for trailing grid blocks).

File: lmcache/v1/compute/attention/metadata.py [hunk 5]
```diff
@@ -83,3 +83,103 @@ def update_from_top_indices(self, top_indices: torch.Tensor):
...
+        top_indices_slice = top_indices[
+            self.sparse_blk_row_size - 1 :: self.sparse_blk_row_size
+        ]
+        top_indices_block = top_indices_slice // self.sparse_blk_col_size
+
+        cols = torch.arange(num_block_col, device=device).expand(
+            len(top_indices_block), -1
+        )
+        mask = cols < top_indices_block.unsqueeze(1)
+        block_mask[: len(top_indices_block)] = mask
```
[the remark this answers is not in the document]
Comment (andyluo7) [reply lmcache/v1/compute/attention/metadata.py:174]: Fixed — `update_from_top_indices` now uses ceiling division for `num_block_row`, and the kernel has an OOB guard for trailing blocks.

File: lmcache/v1/compute/attention/metadata.py [hunk 6]
```diff
@@ -83,3 +83,100 @@ def update_from_top_indices(self, top_indices: torch.Tensor):
...
+        self.is_causal = False
+        device = top_indices.device
+        top_k_num = len(top_indices)
+
+        # Use floor division to match flashinfer behavior — partial
+        # trailing blocks are absorbed into the last full block row,
+        # avoiding empty CSR rows that would cause NaN in the kernel.
+        num_block_row = top_k_num // self.sparse_blk_row_size
+        if num_block_row == 0:
+            num_block_row = 1
```
[the remark this answers is not in the document]
Comment (andyluo7) [reply lmcache/v1/compute/attention/metadata.py]: Fixed — same root cause as the CSR row count issue. Ceiling division + kernel OOB guard covers this.

File: lmcache/v1/compute/attention/triton_kernels/block_sparse_attention.py [hunk 7]
```diff
@@ -0,0 +1,391 @@
...
+    # Load the sparse KV block range for this query block
+    kv_block_start = tl.load(block_indptr_ptr + q_block_idx)
+    kv_block_end = tl.load(block_indptr_ptr + q_block_idx + 1)
+    num_kv_blocks = kv_block_end - kv_block_start
+
+    # Dimension offsets
+    d_offs = tl.arange(0, HEAD_DIM)
+
+    # Load Q block: [BLOCK_M, HEAD_DIM]
+    q_ptrs = Q_ptr + q_offs[:, None] * stride_qm + head_idx * stride_qh + d_offs[None, :] * stride_qd
```
[the remark this answers is not in the document]
Comment (andyluo7) [reply lmcache/v1/compute/attention/triton_kernels/block_sparse_attention.py]: Fixed — both kernels now compute `kv_head_idx = head_idx * NUM_KV_HEADS // NUM_HEADS` and use it for K/V loads. Also added GQA unit tests.

File: lmcache/v1/compute/attention/triton_kernels/block_sparse_attention.py [hunk 8]
```diff
@@ -0,0 +1,386 @@
...
+
+        # Update softmax state
+        m_i = m_new
+        l_i = l_new
+
+    # Final normalization
+    acc = acc / l_i[:, None]
+
+    # Compute LSE = m_i + log(l_i)
+    lse = m_i + tl.log(l_i)
```
[the remark this answers is not in the document]
Comment (andyluo7) [reply lmcache/v1/compute/attention/triton_kernels/block_sparse_attention.py]: Fixed — normalization now uses `tl.where(l_i > 0, acc / l_i, 0.0)` and LSE uses `tl.where(l_i > 0, m_i + log(l_i), -inf)`. Also added `test_empty_mask_returns_zeros` unit test.

File: lmcache/v1/compute/attention/triton_kernels/block_sparse_attention.py [hunk 9]
```diff
@@ -0,0 +1,386 @@
...
+    # Numerically stable: subtract max
+    max_lse = tl.maximum(lse1, lse2)
+    w1 = tl.exp(lse1 - max_lse)
+    w2 = tl.exp(lse2 - max_lse)
+    w_sum = w1 + w2
+
+    # Normalize weights
+    w1 = w1 / w_sum
+    w2 = w2 / w_sum
+
```
[the remark this answers is not in the document]
Comment (andyluo7) [reply lmcache/v1/compute/attention/triton_kernels/block_sparse_attention.py:356]: Fixed — merge kernel now guards against both-inf: `safe_max = where(max_lse == -inf, 0, max_lse)` and `w_sum = where(w_sum > 0, w_sum, 1.0)`. Added `test_both_inf_no_nan` unit test.

File: lmcache/v1/compute/attention/triton_kernels/block_sparse_attention.py [hunk 10]
```diff
@@ -0,0 +1,386 @@
...
+    # Final normalization
+    acc = acc / l_i[:, None]
+
+    # Compute LSE = m_i + log(l_i)
+    lse = m_i + tl.log(l_i)
+
+    # Store output
+    o_ptrs = O_ptr + q_offs[:, None] * stride_om + head_idx * stride_oh + d_offs[None, :] * stride_od
+    tl.store(o_ptrs, acc.to(tl.float16), mask=q_mask[:, None])
```
[the remark this answers is not in the document]
Comment (andyluo7) [reply lmcache/v1/compute/attention/triton_kernels/block_sparse_attention.py]: Fixed — all three kernels now use `Q_ptr.dtype.element_ty` / `O1_ptr.dtype.element_ty` instead of hardcoded `tl.float16`. Unit tests verify both fp16 and bf16 output dtypes.

[the remark this answers is not in the document]
Comment (andyluo7) [reply lmcache/v1/compute/attention/triton_kernels/block_sparse_attention.py]: Fixed — see above, output now uses input tensor dtype.

File: lmcache/v1/compute/attention/triton_sparse.py [hunk 11]
```diff
@@ -0,0 +1,141 @@
...
+    ) -> torch.Tensor:
+        """Compute attention using Triton kernels.
+
+        For causal (full prefill): uses causal_prefill_attention.
+        For non-causal (sparse CacheBlend): uses block_sparse_attention.
+        """
+        from lmcache.v1.compute.attention.triton_kernels import (
+            block_sparse_attention,
+            causal_prefill_attention,
+        )
```
[the remark this answers is not in the document]
Comment (andyluo7) [reply lmcache/v1/compute/attention/triton_sparse.py]: Fixed — `block_sparse_attention` and `causal_prefill_attention` are now imported at module level in `triton_sparse.py`.

File: csrc/event_recorder_hip.h [hunk 12]
```diff
@@ -0,0 +1,79 @@
...
+// CUDA / HIP runtime
+#ifdef USE_ROCM
+  #include <hip/hip_runtime.h>
+using lmcache_stream_t = hipStream_t;
+  #define LMCACHE_LAUNCH_HOST_FUNC hipLaunchHostFunc
+#else
+  #include <hip/hip_runtime.h>
+using lmcache_stream_t = hipStream_t;
+  #define LMCACHE_LAUNCH_HOST_FUNC hipLaunchHostFunc
+#endif
```
[the remark this answers is not in the document]
Comment (andyluo7) [reply csrc/event_recorder_hip.h]: Fixed — `event_recorder.h` now uses proper `#ifdef USE_ROCM` with HIP types in the ROCm branch and CUDA types in the else branch.

File: lmcache/v1/compute/attention/metadata.py [hunk 13]
```diff
@@ -83,3 +83,103 @@ def update_from_top_indices(self, top_indices: torch.Tensor):
...
+            block_indptr,
+        )
+    block_indices = torch.zeros(total_nnz, dtype=torch.int32, device=device)
+    offset = 0
+    for row in range(num_q_blocks):
+        cols = torch.where(block_mask[row])[0].to(torch.int32)
+        n = cols.numel()
+        if n > 0:
+            block_indices[offset : offset + n] = cols
+        offset += n
```
[the remark this answers is not in the document]
Comment (andyluo7) [reply lmcache/v1/compute/attention/metadata.py]: Fixed — `_block_mask_to_csr` now uses vectorized `torch.where(block_mask)[1]` instead of row-by-row Python iteration.

File: lmcache/v1/compute/attention/triton_sparse.py [hunk 14]
```diff
@@ -0,0 +1,140 @@
...
+                query,
+                key,
+                value,
+                block_indices=attn_metadata.block_indices,
+                block_indptr=attn_metadata.block_indptr,
+                sm_scale=self.sm_scale,
+                block_size=attn_metadata.sparse_blk_col_size,
+            )
+            # Store LSE for downstream blending if needed
+            attn_metadata._lse = lse
```
[the remark this answers is not in the document]
Comment (andyluo7) [reply lmcache/v1/compute/attention/triton_sparse.py]: Fixed — now uses the public `attn_metadata.lse` attribute (defined on `LMCTritonSparseMetadata` dataclass) instead of `_lse`.

File: tests/v1/compute/attention/test_triton_sparse_e2e.py [hunk 15]
```diff
@@ -0,0 +1,165 @@
...
+        if t2 < t1 * 0.9:
+            print("✅ CacheBlend speedup observed!")
+        else:
+            print("⚠️  No significant speedup (may need longer prompts)")
+
+        print("\n✅ E2E test completed without errors!")
+
+
+if __name__ == "__main__":
+    main()
```
[the remark this answers is not in the document]
Comment (andyluo7) [reply tests/v1/compute/attention/test_triton_sparse_e2e.py:190]: Added `tests/v1/compute/attention/test_triton_kernels.py` with comprehensive unit tests for all three Triton kernels + the CSR helper. Covers correctness, GQA, dtype preservation, edge cases (empty masks, partial blocks, both-inf merge).

File: lmcache/v1/compute/attention/utils.py [hunk 16]
```diff
@@ -1,14 +1,65 @@
 # SPDX-License-Identifier: Apache-2.0
-# Local
-from .flash_attn import LMCFlashAttnBackend
-from .flash_infer_sparse import LMCFlashInferSparseBackend
+# Standard
+import os
```
[the remark this answers is not in the document]
Comment (andyluo7) [reply lmcache/v1/compute/attention/utils.py]: Fixed — removed.

File: lmcache/v1/compute/attention/triton_sparse.py [hunk 17]
```diff
@@ -0,0 +1,140 @@
...
+                "Call update_from_top_indices() first."
+            )
+            output, lse = block_sparse_attention(
+                query,
+                key,
+                value,
+                block_indices=attn_metadata.block_indices,
+                block_indptr=attn_metadata.block_indptr,
+                sm_scale=self.sm_scale,
+                block_size=attn_metadata.sparse_blk_col_size,
```
[the remark this answers is not in the document]
Comment (andyluo7) [reply lmcache/v1/compute/attention/triton_sparse.py]: This is intentional — `sparse_blk_row_size == sparse_blk_col_size == 32` in the current implementation, so both BLOCK_M and BLOCK_N are the same. Added a comment clarifying this in the code.

File: lmcache/v1/compute/attention/triton_sparse.py [hunk 18]
```diff
@@ -0,0 +1,140 @@
...
+
+class LMCTritonSparseBackend(AttentionInterface):
+    """Triton-based block-sparse attention backend for LMCache.
+
+    Drop-in replacement for LMCFlashInferSparseBackend.
+    Works on both CUDA (via Triton) and ROCm (via Triton).
+    Does NOT require flashinfer.
+    """
+
+    def __init__(self, vllm_attn):
```
[the remark this answers is not in the document]
Comment (andyluo7) [reply lmcache/v1/compute/attention/triton_sparse.py]: Fixed — added type hints to `__init__` and `infer_attn_backend_from_vllm`.

File: lmcache/v1/compute/attention/triton_kernels/block_sparse_attention.py [hunk 19]
```diff
@@ -0,0 +1,386 @@
+"""Triton block-sparse attention kernel for LMCache CacheBlend.
```
[the remark this answers is not in the document]
Comment (andyluo7) [reply lmcache/v1/compute/attention/triton_kernels/block_sparse_attention.py:2]: Fixed — SPDX header added as line 1.

File: test_e2e_rocm.sh [hunk 20]
```diff
@@ -0,0 +1,46 @@
...
+      --tensor-parallel-size 1 \
+      --gpu-memory-utilization 0.9 \
+      --port 8000 \
+      --host 0.0.0.0 \
+      --kv-transfer-config '{\"kv_connector\": \"LMCacheConnector\", \"kv_role\": \"kv_both\"}' \
+      --lmcache-config-file /opt/LMCache/lmcache_config.yaml
+  "
+
+echo "Container started: $CONTAINER_NAME"
+echo "Waiting for server..."
```
[the remark this answers is not in the document]
Comment (andyluo7) [reply test_e2e_rocm.sh]: Fixed — test files moved to `tests/v1/compute/attention/` and hardcoded paths removed.

File: csrc/event_recorder_hip.h (hunk 12 above)
[the remark this answers is not in the document]
Comment (andyluo7) [reply csrc/event_recorder_hip.h]: Fixed — `event_recorder.h` now correctly uses `cuda_runtime.h` / `cudaStream_t` / `cudaLaunchHostFunc` in the non-ROCm branch.

File: tests/v1/compute/attention/test_triton_sparse_e2e.py [hunk 21]
```diff
@@ -0,0 +1,184 @@
...
+from dataclasses import asdict
+import contextlib
+import os
+import time
+
+# Third Party
+from transformers import AutoTokenizer
+from vllm import LLM, SamplingParams
+from vllm.config import KVTransferConfig
+from vllm.engine.arg_utils import EngineArgs
```
Comment (ApostaC) [inline tests/v1/compute/attention/test_triton_sparse_e2e.py]: We don't have vLLM in UT environments, so this will cause UT to fail.

**Review by ApostaC — COMMENTED** [review]
Quick small catch on the UT. PTAL.

File: tests/v1/compute/attention/test_triton_sparse_e2e.py (hunk 21 above)
Comment (andyluo7) [reply tests/v1/compute/attention/test_triton_sparse_e2e.py]: Good catch — added `pytest.importorskip("vllm")` so the E2E test is skipped gracefully when vLLM is not installed. The unit tests in `test_triton_kernels.py` don't depend on vLLM at all.

**Review by deng451e — APPROVED** [review]
LGTM!
